frontends/common/constantFolding.cpp

```diff
@@ -710,9 +710,18 @@ const IR::Node *DoConstantFolding::postorder(IR::LNot *e) {
 }
 
 const IR::Node *DoConstantFolding::postorder(IR::Mux *e) {
-    if (!typesKnown)
-        // We want the typechecker to look at the expression first
-        return e;
+    if (!typesKnown) {
+        // Action enum switch case labels must be action names.
+        // If we are in the context of a switch label, we want the typechecker
+        // to look at the label expression first.
+        if (const auto *switchStatement = findContext<IR::SwitchStatement>()) {
```

Comment:
I wonder how expensive this search could be for deeply nested expressions...
In the worst case it will traverse to the root for each mux expression. 

Maybe https://github.com/p4lang/p4c/issues/3622 should be valid code instead?

Reply:
I thought that #3633 was added because:

1. P4 spec says that action enum switch labels can only be action names, and #3633 prevents them from being mux expressions that evaluate to action names at compile-time
2. P4 spec says that mux cannot act on action decls at all

I have not looked at the spec to verify (1) or (2). I am just going by what was stated in https://github.com/p4lang/p4c/issues/3622#issue-1423167785.

I would be fine with reverting #3633 if you don't think we actually need the error message added by it.

Reply:
Yeah I have been thinking about changing the specification. I wonder whether there are any downsides in having compile-time-known ternary operator support for labels. 

I just did some quick performance measurements and this change does not cause significant degradation for me, which is good. 

Would be great to have an approach that does not require searching for the context for every Mux, but I can not think of one right now.

Reply:
Maybe we should just revert #3622 and reopen #3633 until we decide whether to change the spec or not? I also cannot think of any downside in having compile-time-known ternary operator support for labels.

Reply:
> I wonder how expensive this search could be for deeply nested expressions...
> In the worst case it will traverse to the root for each mux expression.

I really do not think it is a problem. There are lots of much more severe overheads in other places, so this is just noise. We can certainly think of speeding-up `findContext` in general :)

Reply:
Well, it can have some impacts for analysis tools that use constant folding on complex expressions. In the case of the tool I am using, it looks like it causes a ~6% slowdown on larger programs. But I understand that this is a non-standard use of the pass. 

My concern is that this can be a death-by-a-thousand-paper-cuts type of situation. The transform pass can definitely be improved to be faster etc but we are also some inefficient lookups here.

Reply:
> My concern is that this can be a death-by-a-thousand-paper-cuts type of situation.

And we are already here actually :)

> The transform pass can definitely be improved to be faster etc but we are also some inefficient lookups here.

+1

Reply:
@fruffy Hopefully the recent changes will eliminate your concerns about the performance.

Reply:
This is a good solution, thanks!